debug_reg_printer.c

```diff
@@ -61,11 +61,13 @@ static uint64_t riscv_debug_reg_field_value(riscv_debug_reg_field_info_t field,
 }
 
 static unsigned int riscv_debug_reg_fields_to_s(char *buf, unsigned int offset,
-		riscv_debug_reg_field_list_t list, riscv_debug_reg_ctx_t context, uint64_t value)
+	struct riscv_debug_reg_field_list_t (*get_next)(riscv_debug_reg_ctx_t contex),
+	riscv_debug_reg_ctx_t context, uint64_t value)
 {
 	unsigned int curr = offset;
 	curr += get_len_or_sprintf(buf, curr, " { ");
-	for (; list.get_next; list = list.get_next(context)) {
+	for (struct riscv_debug_reg_field_list_t list; get_next; get_next = list.get_next) {
+		list = get_next(context);
```

Comment:
I see how the old code would omit the last field, because when list.get_next is NULL you still need to print out one more field.